runtime/minor_gc.c

```diff
@@ -682,6 +682,15 @@ void caml_empty_minor_heap_promote (struct domain* domain, int participating_cou
   domain_state->stat_minor_collections++;
   domain_state->stat_promoted_words += domain_state->allocated_words - prev_alloc_words;
 
+#ifdef DEBUG
+  caml_global_barrier();
```

Comment:
I'm a bit worried about changes in behaviour by having barriers when in debug mode. 
Can we move this block into `caml_empty_minor_heap_domain_clear` (which is called after the `EV_MINOR_LEAVE_BARRIER`) and do away with the barrier in debug mode?

Reply:
That was my initial plan but I couldn't satisfy myself there weren't any references in to the minor heap from other domains at that point - as there isn't a barrier.

Though the fact there could still be references (e.g from the finalisers) is a bug in itself?

Reply:
So `caml_empty_minor_heap_domain_clear` is called on the other side of the minor GC leave barrier (see `EV_MINOR_LEAVE_BARRIER`) with the opportunistic work. Anything on the other side of that barrier can't reference data in any minor heap and is a bug. 

`caml_empty_minor_heap_domain_clear` currently scrubs the remembered set structures, updates which finalisers might be young, etc. So I think it is the right place to do more debug scrubbing.

Reply:
Ok, that makes sense. Will move it over there.